src/mc-text-search-str-encode.c

```diff
@@ -0,0 +1,225 @@
+                                            uint32_t unfolded_codepoint_len,
+                                            const mc_FLE2PrefixInsertSpec_t *spec) {
+    return generate_prefix_or_suffix_tree(base_str, unfolded_codepoint_len, spec->lb, spec->ub, true);
+}
+
+static uint32_t calc_number_of_substrings(uint32_t strlen, uint32_t lb, uint32_t ub) {
+    // There are len - i + 1 substrings of length i in a length len string.
+    // Therefore, the total number of substrings with length between lb and ub
+    // is the sum of the integers inclusive between A = len - ub + 1 and B = len - lb + 1,
+    // A <= B. This has a closed form: (A + B)(B - A + 1)/2.
+    if (lb > strlen) {
+        return 0;
+    }
+    uint32_t largest_substr = BSON_MIN(strlen, ub);
+    uint32_t largest_substr_count = strlen - largest_substr + 1;
+    uint32_t smallest_substr_count = strlen - lb + 1;
+    return (largest_substr_count + smallest_substr_count) * (smallest_substr_count - largest_substr_count + 1) / 2;
+}
+
+static mc_substring_set_t *generate_substring_tree(const mc_utf8_string_with_bad_char_t *base_str,
+                                                   uint32_t unfolded_codepoint_len,
+                                                   const mc_FLE2SubstringInsertSpec_t *spec) {
+    // 16 * ceil(unfolded len / 16)
+    uint32_t cbclen = 16 * (uint32_t)((unfolded_codepoint_len + 15) / 16);
+    if (unfolded_codepoint_len > spec->mlen || cbclen < spec->lb) {
+        // No valid substrings, return empty tree
+        return NULL;
+    }
+
+    // If you are following along with the OST paper, a slightly different calculation of msize is used. The following
+    // justifies why that calculation and this calculation are equivalent.
+    // At this point, it is established that:
+    //     beta <= mlen
+    //     lb <= cbclen
+    //     lb <= ub <= mlen
+    //
+    // So, the following formula for msize in the OST paper:
+    //     maxkgram_1 = sum_(j=lb, ub, (mlen - j + 1))
+    //     maxkgram_2 = sum_(j=lb, min(ub, cbclen), (cbclen - j + 1))
+    //     msize      = min(maxkgram_1, maxkgram_2)
+    // can be simplified to:
+    //     msize      = sum_(j=lb, min(ub, cbclen), (min(mlen, cbclen) - j + 1))
+    //
+    // because if cbclen <= ub, then it follows that cbclen <= ub <= mlen, and so
+    //     maxkgram_1 = sum_(j=lb, ub, (mlen - j + 1))          # as above
+    //     maxkgram_2 = sum_(j=lb, cbclen, (cbclen - j + 1))    # less or equal to maxkgram_1
+    //     msize      = maxkgram_2
+    // and if cbclen > ub, then it follows that:
+    //     maxkgram_1 = sum_(j=lb, ub, (mlen - j + 1))          # as above
+    //     maxkgram_2 = sum_(j=lb, ub, (cbclen - j + 1))        # same sum bounds as maxkgram_1
+    //     msize      = sum_(j=lb, ub, (min(mlen, cbclen) - j + 1))
+    // in both cases, msize can be rewritten as:
+    //     msize      = sum_(j=lb, min(ub, cbclen), (min(mlen, cbclen) - j + 1))
```

Comment:
On second thought, maybe let's just keep this simple and calculate it the way it's done in the paper?
```
uint32_t maxkgram_1 = calc_number_of_substrings(mlen, spec->lb, spec->ub);
uint32_t maxkgram_2 = calc_number_of_substrings(cbclen, spec->lb, BSON_MIN(spec->ub, cbclen));
uint32_t msize = BSON_MIN(maxkgram_1, maxkgram_2);
```

Reply:
Eh, I don't like doing more calculations just in order to match the paper. I feel that the variable names and the inline comments give enough reason for why this counting method is correct -- it's clear that we are padding the length to the lesser of cbclen or mlen, and then calculating how many substrings between length lb and ub there would be if the string actually was that max padded length.